File: JavaScript/Medium/ValidateBinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
 var isValidBST = function(root) {
    
    let valid = function(node, left, right) {
        if (node === null) return true;
        
        if (node.val <= left || node.val >= right) return false;
        
        return valid(node.left, left, node.val) && valid(node.right, node.val, right);
    }
  
    return valid(root, Math.MIN_SAFE_NUMBER, Math.MAX_SAFE_NUMBER)
};

/*

    DFS return false if out of range
    
                5   
            |       |
            1       6
                 |     |
                 3     7         
                 
    -MAX < 5 < MAX
    -MAX < 1 < 5
       5 < 6 < MAX  
       5 < 3 < 6    --- false
    
    Time: O(n)
    Space: O(h)
                 
*/

File: C#/Medium/ValidateBinarySearchTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root)
    {
        return IsValidBST(root, long.MinValue, long.MaxValue);
    }

    private bool IsValidBST(TreeNode root, long min, long max)
    {
        if (root == null) return true;
        if (root.val <= min || root.val >= max) return false;

        if (!IsValidBST(root.left, min, root.val) || !IsValidBST(root.right, root.val, max)) return false;

        return true;
    }
}

/*

    DFS - Keep a min and max range that gets updates as the BST is traversed.

    Time: O(n)
    Space: O(h)

    Where n is the number of nodes in the tree and h is the height                

*/
